crates/modelardb_compression/src/models/swing.rs

```diff
@@ -326,6 +350,26 @@ fn compute_slope_and_intercept(
     }
 }
 
+/// Computes numerator and denominator for Equation 6 
+/// in Section 3.2 of [Swing and Slide paper](https://dl.acm.org/doi/10.14778/1687627.1687645).
+/// It is a part of recording mechanism for finding the optimal slope with minimum MSE.
+fn compute_local_optimal_slope(
+    start_time: Timestamp,
+    first_value: f64,
+    end_time: Timestamp,
+    last_value: f64,
+) -> (f64, f64) {
+    if first_value == last_value || (first_value.is_nan() || last_value.is_nan()) {
+        (0.0, 0.0)
+    } else {
+        debug_assert!(first_value.is_finite(), "First value is not finite.");
+        debug_assert!(last_value.is_finite(), "Last value is not finite.");
+        let numerator = (last_value - first_value) * (end_time - start_time) as f64;
+        let denominator = ((end_time - start_time) as f64).powi(2);
+        (numerator, denominator)
+    }
+}
+
 #[cfg(test)]
 mod tests {
```

Comment:
Unit tests should be added to verify that the code added in this PR makes Swing produce linear functions with minimum MSE. The unit tests should test all aspects of the new code, e.g., by compression sequences of data points that Swing would previously not fit a linear function to with an optimal MSE. If possibly, tests that use data points generated using `proptest` or `modelardb_common::test::data_generation` should be added to improve our confidence that the code is correct and to catch bugs that happen due to changes in the code elsewhere. Remember to check that the tests actually tests the effect of the code added by this PR, e.g., that the tests fail if `compute_local_optimal_slope()` is not called.

Reply:
> "e.g., by compression sequences of data points that Swing would previously not fit a linear function to with an optimal MSE " the length of approximated sequences doesn't increase as a result of this change.

Yes the length should be the same but the MSE should improve as a result of these changes. Thus, it would be good if tests were added that verify that the models that are now being produced now has a lower MSE. For example, by using a sequence of data points that Swing compressed with a linear function with a higher MSE before these changes were added so the test fails without these changes. The best case would be to automatically generate random data and then check that the MSE is always smaller than before, however, I do not know if that is possible without keeping the old version as well. Thus, using a sequence of data points that now has a reduced MSE is probably the more reasonable approach despite not testing the code as extensively.